### PR TITLE
[FIX](datatype) Implement hash func with array/map/struct type

### PR DESCRIPTION
**Pull request by amorynan**
## Proposed changes

Issue Number: close #xxx
we do not Implement any hash functions in array/map/struct column  , so we use sql like this will make be core
```
select * from (
        select
            bdp.nc_num,
            collect_list(distinct(bd.catalog_name)) as catalog_name,
            material_qty
        from
            dataease.bu_delivery_product bdp
            left join dataease.bu_trans_transfer btt on bdp.delivery_product_id = btt.delivery_product_id
            left join dataease.bu_delivery bd on bdp.delivery_id = bd.delivery_id
        where
            bd.val_status in ('10', '20', '30', '90')
            and bd.delivery_type in (0, 1, 2)
        group by
            nc_num,
            material_qty
        union
        ALL
        select
            bdp.nc_num,
            collect_list(distinct(bd.catalog_name)) as catalog_name,
            material_qty
        from
            dataease.bu_trans_transfer btt
            left join dataease.bu_delivery_product bdp on bdp.delivery_product_id = btt.delivery_product_id
            left join dataease.bu_delivery bd on bdp.delivery_id = bd.delivery_id
        where
            bd.val_status in ('10', '20', '30', '90')
            and bd.delivery_type in (0, 1, 2)
        group by
            nc_num,
            material_qty
) aa;
```

core :
```
F0626 13:15:33.174283 31087 column.h:360] update_hashes_with_value xxhash not supported
*** Check failure stack trace: ***
    @     0x55dc63769eed  google::LogMessage::Fail()
    @     0x55dc6376c429  google::LogMessage::SendToLog()
    @     0x55dc63769a56  google::LogMessage::Flush()
    @     0x55dc6376ca99  google::LogMessageFatal::~LogMessageFatal()
    @     0x55dc5da9d1a0  doris::vectorized::IColumn::update_hashes_with_value()
    @     0x55dc630ea068  doris::vectorized::VDataStreamSender::send()
    @     0x55dc5e340e89  doris::PlanFragmentExecutor::open_vectorized_internal()
    @     0x55dc5e341dda  doris::PlanFragmentExecutor::open()
    @     0x55dc5e319dae  doris::FragmentExecState::execute()
    @     0x55dc5e31d1a5  doris::FragmentMgr::_exec_actual()
    @     0x55dc5e31d8ca  _ZNSt17_Function_handlerIFvvEZN5doris11FragmentMgr18exec_plan_fragmentERKNS1_23TExecPlanFragmentParamsESt8functionIFvPNS1_20PlanFragmentExecutorEEEEUlvE_E9_M_invokeERKSt9_Any_data
    @     0x55dc5e5d3b15  doris::ThreadPool::dispatch_thread()
    @     0x55dc5e5c924f  doris::Thread::supervise_thread()
    @     0x7fd2f4e07ea5  start_thread
    @     0x7fd2f52feb0d  __clone
    @              (nil)  (unknown)
*** Query id: 41bbb29cba364043-881fb7e8e94c3ff0 ***
*** Aborted at 1687756533 (unix time) try "date -d @1687756533" if you are using GNU date ***
*** Current BE git commitID: Unknown ***
*** SIGABRT unkown detail explain (@0x3e900005f30) received by PID 24368 (TID 0x7fd2193f0700) from PID 24368; stack trace: ***
 0# doris::signal::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*) at /root/doris/be/src/common/signal_handler.h:420
 1# 0x00007FD2F5236400 in /lib64/libc.so.6
 2# gsignal in /lib64/libc.so.6
 3# abort in /lib64/libc.so.6
 4# 0x000055DC637748D9 in /data/opt/module/doris/be/lib/doris_be
 5# 0x000055DC63769EED in /data/opt/module/doris/be/lib/doris_be
 6# google::LogMessage::SendToLog() in /data/opt/module/doris/be/lib/doris_be
 7# google::LogMessage::Flush() in /data/opt/module/doris/be/lib/doris_be
 8# google::LogMessageFatal::~LogMessageFatal() in /data/opt/module/doris/be/lib/doris_be
 9# doris::vectorized::IColumn::update_hashes_with_value(unsigned long*, unsigned char const*) const at /root/doris/be/src/vec/columns/column.h:360
10# doris::vectorized::VDataStreamSender::send(doris::RuntimeState*, doris::vectorized::Block*) at /root/doris/be/src/vec/sink/vdata_stream_sender.cpp:581
11# doris::PlanFragmentExecutor::open_vectorized_internal() at /root/doris/be/src/runtime/plan_fragment_executor.cpp:322
12# doris::PlanFragmentExecutor::open() at /root/doris/be/src/runtime/plan_fragment_executor.cpp:261
13# doris::FragmentExecState::execute() at /root/doris/be/src/runtime/fragment_mgr.cpp:261
14# doris::FragmentMgr::_exec_actual(std::shared_ptr<doris::FragmentExecState>, std::function<void (doris::PlanFragmentExecutor*)>) at /root/doris/be/src/runtime/fragment_mgr.cpp:508
15# std::_Function_handler<void (), doris::FragmentMgr::exec_plan_fragment(doris::TExecPlanFragmentParams const&, std::function<void (doris::PlanFragmentExecutor*)>)::{lambda()#1}>::_M_invoke(std::_Any_data const&) at /var/local/ldb-toolchain/include/c++/11/bits/std_function.h:291
16# doris::ThreadPool::dispatch_thread() at /root/doris/be/src/util/threadpool.cpp:543
17# doris::Thread::supervise_thread(void*) at /root/doris/be/src/util/thread.cpp:455
18# start_thread in /lib64/libpthread.so.0
19# clone in /lib64/libc.so.6 
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by amorynan** [comment]
run buildall

**Review by jacktengg — APPROVED** [review]
LGTM

File: be/src/vec/sink/vdata_stream_sender.cpp [hunk 1]
```diff
@@ -629,7 +630,8 @@ Status VDataStreamSender::send(RuntimeState* state, Block* block, bool eos) {
                 std::vector<SipHash> siphashs(rows);
                 // result[j] means column index, i means rows index
                 for (int j = 0; j < result_size; ++j) {
-                    block->get_by_position(result[j]).column->update_hashes_with_value(siphashs);
+                    unpack_if_const(block->get_by_position(result[j]).column)
```
Comment (yiguolei) [inline be/src/vec/sink/vdata_stream_sender.cpp:633]: add some comment here, to explain whey do we need uppack const column.

File: be/src/vec/columns/column_array.cpp [hunk 2]
```diff
@@ -271,6 +271,69 @@ void ColumnArray::update_hash_with_value(size_t n, SipHash& hash) const {
...
 }
 
+void ColumnArray::update_hashes_with_value(std::vector<SipHash>& hashes,
+                                           const uint8_t* __restrict null_data) const {
+    SIP_HASHES_FUNCTION_COLUMN_IMPL();
+}
+
+// for every array row calculate xxHash
+void ColumnArray::update_xxHash_with_value(size_t n, uint64_t& hash) const {
+    size_t elem_size = size_at(n);
```
Comment (yiguolei) [inline be/src/vec/columns/column_array.cpp:281]: 我感觉这个逻辑不对啊，比如 [[1,1,1], [2,2,2,2], [3]] 这个带nested的array

File: be/src/vec/sink/vdata_stream_sender.cpp (hunk 1 above)
Comment (amorynan) [reply be/src/vec/sink/vdata_stream_sender.cpp:634]: done!

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

**Comment by amorynan** [comment]
run arm

**Comment by amorynan** [comment]
run arm

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**